lib/beaker/hypervisor/aws_sdk.rb

```diff
@@ -220,125 +221,171 @@ def kill_zombie_volumes
-      end
+      msg = "aws-sdk: launching %p on %p using %p/%p%s" %
+            [host.name, amitype, amisize, image_type,
+             subnet_id ? ("in %p" % subnet_id) : '']
+      @logger.notify(msg)
+      config = {
+        :count => 1,
+        :image_id => image_id,
+        :monitoring_enabled => true,
+        :key_pair => ensure_key_pair(region),
+        :security_groups => [security_group],
+        :instance_type => amisize,
+        :disable_api_termination => false,
+        :instance_initiated_shutdown_behavior => "terminate",
+        :block_device_mappings => block_device_mappings,
+        :subnet => subnet_id,
+      }
+      region.instances.create(config)
+    end
 
+    # Create EC2 instances for all hosts, tag them, and wait until
+    # they're running.
+    #
+    # @return [void]
+    # @api private
+    def launch_all_nodes
+      @logger.notify("aws-sdk: launch all hosts in configuration")
+      ami_spec = YAML.load_file(@options[:ec2_yaml])["AMI"]
+      global_subnet_id = @options['subnet_id']
+      global_subnets = @options['subnet_ids']
+      if global_subnet_id && global_subnets
+        raise RuntimeError, 'Config specifies both subnet_id and subnet_ids'
+      end
+      any_subnet_hosts, other_hosts = @hosts.partition do |host|
+        global_subnets && ! (global_subnet_id || host['subnet_id'])
+      end
+      instances = [] # Each element is {:instance => i, :host => h}
+      begin
+        @logger.notify("aws-sdk: launch instances not particular about subnet")
+        # Shuffle the subnets so we don't always hit the same one
+        # first, and cycle though the subnets independently of the
+        # host, so we stick with one that's working.  Try each subnet
+        # once per-host.
+        subnet_i = 0
+        if global_subnets
+          shuffnets = global_subnets.shuffle
+          any_subnet_hosts.each do |host|
+            instance = nil
+            shuffnets.length.times do
+              begin
+                subnet_id = shuffnets[subnet_i]
+                instance = create_instance(host, ami_spec, subnet_id)
+                instances.push({:instance => instance, :host => host})
+                break
+              rescue AWS::EC2::Errors::InsufficientInstanceCapacity => ex
+                @logger.notify("aws-sdk: hit #{subnet_id} capacity limit; moving on")
+                subnet_i = (subnet_i + 1) % shuffnets.length
+              end
+            end
+            if instance.nil?
+              raise RuntimeError, "unable to launch host in any subnet"
+            end
+          end
+        end
+        other_hosts.each do |host|
+          subnet_id = host['subnet_id'] || global_subnet_id || nil
+          instance = create_instance(host, ami_spec, subnet_id)
+          instances.push({:instance => instance, :host => host})
+        end
+        wait_for_status(:running,
```

Comment:
@kbarber I don't disagree, but wasn't sure at the time (with respect to Ruby) it'd be palatable to have it accept the somewhat internal sequence of {:instance, :host} values instead.  Would that be preferable?